assembly/src/tokens/lines.rs

```diff
@@ -0,0 +1,502 @@
+                None
+            }
+        }
+    }
+}
+
+// LINE INFO
+// ================================================================================================
+
+/// A processed line with source location.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct LineInfo<'a> {
+    contents: Option<&'a str>,
+    docs: Vec<&'a str>,
+    line_number: u32,
+    char_offset: u32,
+}
+
+impl<'a> LineInfo<'a> {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Create a new instance of [LineInfo] with the provided line number and first non-blank char
+    /// offset.
+    pub fn new(line_number: u32, char_offset: u32) -> Self {
+        Self {
+            contents: None,
+            docs: Vec::new(),
+            line_number,
+            char_offset,
+        }
+    }
+
+    /// Replaces the doc comments with the provided argument.
+    pub fn with_docs<I>(mut self, docs: I) -> Self
+    where
+        I: IntoIterator<Item = &'a str>,
+    {
+        self.docs = docs.into_iter().collect();
+        self
+    }
+
+    /// Replaces the line comments with the provided argument.
+    pub fn with_contents(mut self, contents: &'a str) -> Self {
+        self.contents.replace(contents.trim_end());
+        self
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the line contents, if present.
+    ///
+    /// # Examples
+    ///
+    /// ```masm
+    /// #! some doc comment
+    /// # some line comment
+    ///
+    /// add mul
+    /// ```
+    ///
+    /// `add mul` is returned.
+    pub const fn contents(&self) -> Option<&'a str> {
+        self.contents
+    }
+
+    /// Returns the doc comments.
+    ///
+    /// # Examples
+    ///
+    /// ```masm
+    /// #!      doc comments
+    /// #! for foo procedure          
+    /// #! with examples
+    /// export.foo
+    ///   add
+    /// end
+    /// ```
+    ///
+    /// `["doc comments", "for foo procedure", "with examples"]` is returned.
+    pub fn docs(&self) -> &[&'a str] {
+        &self.docs
+    }
+
+    /// Returns the line number, starting at `1`.
+    ///
+    /// # Examples
+    ///
+    /// ```masm
+    /// #! doc comments
+    /// export.foo
+    ///   add
+    /// end
+    /// ```
+    ///
+    /// `add` will return `3`.
+    pub const fn line_number(&self) -> u32 {
+        self.line_number
+    }
+
+    /// Returns the first non-whitespace character offset.
+    ///
+    /// # Examples
+    ///
+    /// ```masm
+    ///   add mul
+    /// ```
+    ///
+    /// `2` is returned.
+    pub const fn _char_offset(&self) -> u32 {
+        self.char_offset
+    }
+}
+
+// UNIT TESTS HELPERS
+// ================================================================================================
```

Comment:
nit: we usually wrap tests and test related functions into a `tests` module like so:
```Rust
#[cfg(test)]
mod tests {
    ...
}

```

Reply:
Fixed; was done like that to avoid the extra indentation